### ISSUE TITLE
Sales disappearing from system/ Tax code issue

### ISSUE THREAD

**Issue by esbyrt** [opensourcepos/opensourcepos#1635]
Issue / Question / Bug

Before submitting an issue please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

I'm reporting an issue of an unmodified OSPOS installation
Installation information

OSPOS version is: 3.1.1
OSPOS git commit hash is:
PHP version is: (e.g. 5.5, 5.6, 7.0, 7.1) 7.1
MySQL or MariaDB version is: (e.g. MySQL 5.5, MySQL 5.6, MySQL 5.7, MariaDB 10.0, MariaDB 10.1, MariaDB 10.2) 10.0 .32
OS and version is: (e.g. CentOS 6.9, Ubuntu 16.4, Windows 10) Windows 10/Windows 8
WebServer is: (e.g. Apache 2.2, Apache 2.4, Nginx 1.12, Nginx 1.13) Apache 2.4.27

Issue / Question / Bug

### Expected behaviour

Enter a sale, print it out, sale should show on daily report

### Actual behaviour
Enter a sale, nothing happens when clicking print (no error message but nothing prints either). Sale does not show on daily report. When I check in phpmyadmin I see the pos sale numbers jump from 283 to 289. There is no way the users could get into the backend to delete invoices and they wouldn't know how to delete them in the POS. Has anyone else encountered this? If the POS sale number increments up shouldn't the sale be there somewhere?

I didn't see a table that tracks deleted invoices. Is there a mechanism for that?

**Comment by daN4cat** [comment]
Are you running everything on a local machine running Windows?

**Comment by esbyrt** [comment]
No it's running on a sub domain on their website. They connect via wifi in the store. I considered a lost wifi connection mid sale but.... this is the third time this has happened since the update to 3.1.1 and it never happened before. The first time I wrote off to user error. Now I don't think that's it.

**Comment by SteveIreland** [comment]
Do you create quotes or suspend sales that might be discarded?  Those will cause gaps in the sale_id numbers in the 3.1.1 release.  In the 3.2 release that has been changed so that they are instead flagged as canceled.

**Comment by SteveIreland** [comment]
I'm not aware of any practical limitation.  The primary key is by sale_id, item_id,  and line number so there isn't a practical limit.  I did notice that the three digit line number is being incremented for each line, so one could say that there is a presumed limit of 999 lines per invoice.  But I don't think that would actually be the case since the line number could wrap and the system would rarely care.  Technically the key structure would suggest that the line number primarily serves the purpose of allowing the same item to be added multiple times with a different serial number.    However, at the moment it seems to actually be used to track the sequence in which lines are added.

I know this is off topic of the thread but I thought it needed to be mentioned and the design intent be established so that an error (if there is one) can be addressed.

**Comment by daN4cat** [comment]
@esbyrt so far nobody reported such issue but you might have a corner case and you need to debug for us to reproduce and fix.
I assume @SteveIreland has a live setup of the latest so he should get this soon or later too if it's a real issue.

**Comment by esbyrt** [comment]
The user tried again to enter the sale and once again it printed weird, more like a screen shot than a sales receipt (screen shot attached) and didn't enter into the system. Other sales went through ok on the other computer so I'm thinking at this point the problem might be with her computer. I will go on site Monday or Tuesday to troubleshoot and let you know what I find.
![img_5366](https://user-images.githubusercontent.com/28318020/30787606-09ca4616-a14a-11e7-8077-7cde94e855f6.JPG)

**Comment by daN4cat** [comment]
Do you know how many sales they have done so far? And how many items per transaction?

**Comment by daN4cat** [comment]
BTW it is a screen shot. I bet that those people are not completing the transaction properly. Ask them to show you what they do.

**Comment by esbyrt** [comment]
Total sales to date are about 300. Probably less than 25 since I updated to the new release. This particular transaction has a lot of items since it's a summary of cash sales at a trade show. However, troubleshooting this will have to wait until next week. The store owners are off to another trade show and the temp clerk is doing manual sales receipts this week. As soon as I can get into the store to see what they are doing I will update here. Thanks!

**Comment by daN4cat** [comment]
@esbyrt if you have some time give a try adding many items and see if it fails. So you can help with the debug and test. Thanks.

**Comment by esbyrt** [comment]
Hi all. I have done some testing using as many of the same items in the failed sale as I could make out. Initially I entered 11 items for a cash sale, applied the payment, clicked complete. The browser screen went blank and the address at the top was mydomain/opensourcepos/public/sales/complete.  Going back into ospos the sale is still on the screen but clicking complete again has the same behaviour. The only way out is to cancel the sale. I tried again with the first 4 items. That worked. I tried the next 5. Failed. Thinking there might be something wrong with one of the items I tried each alone as a sale. Each worked fine with a 1 item sale. Tried again with 3 of them. Failed. The problem doesn't seem to be related to number of items or a particular item. For some reason the "complete" is failing to complete. Any ideas on where to look next?

**Comment by daN4cat** [comment]
Also if you could put the failing items in a csv to import I can test next days.

**Comment by esbyrt** [comment]
Here's the csv of the failing items.
[failed items.zip](https://github.com/opensourcepos/opensourcepos/files/1335397/failed.items.zip)

**Comment by objecttothis** [comment]
This might have nothing to do with it but on a very old version of OSPOS we had an issue of the session expiring during long sales and losing the sale and having to start over.  We worked around this by changing $config['sess_expiration'] = 7200; to 0. It doesn't sound to me like quite the same issue though.

**Comment by daN4cat** [comment]
I would need to find the time to check this.

**Comment by esbyrt** [comment]
As posted on my missing taxes issue - daN4cat I honestly don't remember for sure if I ran the migration when I did the upgrade. I remember seeing that the migrate button would show up if it was needed so wouldn't have been too worried if it didn't. Anyway, I just backed up the database and ran migrate. Went to do an invoice with 7 random items and applied payment but clicking complete brings me to a blank screen instead of the invoice and the sale fails to enter into the system.

At this point my users are understandably getting frustrated. Version 3.0.2 had an issue with gift cards but worked for them otherwise. I am thinking i might have to roll back to the previous version. Can someone give me some guidance on how to do that? I do have a backup of my previous ospos folder on the server. Could I just revert the folder name to opensourcepos and have it work with the migrated databases?
Thanks

**Comment by daN4cat** [comment]
Does that happen only with invoice or normal receipt too?

**Comment by daN4cat** [comment]
Also you could try on a copy of the database to take the latest code as it's pretty stable and a pre 3.2.0 release. We are holding it back to solve a translation issue and maybe add a couple of new features.
See if the latest gives the same result.

**Comment by daN4cat** [comment]
Use bintray zip file. No need to run any upgrade scripts as it's all done automatically now on first login.

**Comment by esbyrt** [comment]
Hi. It's actually on receipts that is happening. They don't use invoicing. I keep calling it an invoice because it prints on full size paper. Yes I will try the latest code but I find this site so darn confusing trying to find updates. Can you link me to the file please? Thanks so much.

**Comment by odiea** [comment]
https://bintray.com/jekkos/opensourcepos/download_file?file_path=opensourcepos.20171021130858.master.2b2396.zip

**Comment by esbyrt** [comment]
Okay I have 3.2.0 installed as per the directions except I didn't run any upgrade script as  you said it was automatic now. When I login the opening screen is for 3.2 but there is no migrate to click on like the instructions say. I checked my user permissions and I have access to everything. Should I run the upgrade script after all?

**Comment by esbyrt** [comment]
Still wondering if I need to run a migrate or update script of some kind. My sales not completing issue remains. I tried numerous items with one customer. 7 items failed always. 3 items failed sometimes. No rhyme or reason, an item will fail once then work the next time. One thing I did see when I hit the back button on the browser and refresh the error "Amount tendered must be a number" is at the top of the screen.  This error was there whether I selected cash, debit or credit as the payment type.

**Comment by daN4cat** [comment]
No need to run anything as any upgrade is done automatically, even the migrate.

**Comment by daN4cat** [comment]
Have you tried to create 10 simple items with names like item 1, item 2, etc? I want to exclude the item names. Also I assume you use English as language. Last thing, please disable send statistics from config page. Give a try and let me know if using this items it's fine or not.

**Comment by esbyrt** [comment]
Hi. I created a bunch of generic items as suggested and they always went through ok. I then tried to find what was different about the items that wouldn't go through. I found that there are two Tax 1 codes that say Tax 1 but have different tax categories. One Tax 1 is category 1 the other tax 1 is category 0. On a sale where two items have these two different codes the complete fails. 
![fail3](https://user-images.githubusercontent.com/28318020/31976693-30119ac2-b8f6-11e7-9636-5a5d301aa604.PNG)
So now that I know where the glitch is can I update all the items tax category to fix the problem? What tax category SHOULD be there? I noticed even recently added consecutive items ended up with different tax categories. When adding an item the default GST shows up automatically so I don't know how any item would end up different. Thoughts??

**Comment by daN4cat** [comment]
@SteveIreland can you help? You added the tax category part and know more.

**Comment by daN4cat** [comment]
I use VAT with tax inclusive, so I can't help with this.
@SteveIreland introduced that new tax system, so he's the best person to look into this.

**Comment by esbyrt** [comment]
After messing around with lots of settings I enabled Customer Origin Tax Code in store config. That seems to have worked around the problem. I tested with a sale with a few items that have been showing two tax lines up to now. They now show as only one tax line and the sale completes properly. I can't quite follow the logic through on the various tax settings but this sure looks like a bug to me.

**Comment by SteveIreland** [comment]
I’ll take a look at this over the weekend

**Comment by daN4cat** [comment]
@SteveIreland any news about this issue?

**Comment by SteveIreland** [comment]
I'm running through a series of tests this morning to determine if there is a problem or if there is just some confusion.  OSPOS doesn't support GST taxation unless you've slipped something in while I wasn't looking.

As it is right now there are 3 tax configurations ... one is VAT ("Tax Included" checked), another is Customer Sales Tax ("Customer Sales Tax Support" checked), and then we have the original Simple Sales Tax (neither is checked).

The tax configuration needs to be established right from the start and cannot be changed .... that I know for a fact.  But after my tests I'll have a better idea.

**Comment by SteveIreland** [comment]
Tax category 0 isn't a valid tax category  The system pre-loads the tax categories with 1 for standard, 2 for service, and 3 for alcohol.

@esbyrt The only way the screen print  you included on this task was if one was done while in "VAT" tax mode and the other was added while in Sales tax mode.  The modes should not be mixed.

However, if the sale was done totally while in Sales Tax mode then you must have tried out a feature I included in anticipation of future development but doesn't necessarily make sense at this point.  In fact I was thinking that it might provide an opening for me to add support for GST but it's not there yet.

Make sure all of  your tax categories are flagged as Sales Tax and not as VAT Tax.

**Comment by SteveIreland** [comment]
@jekkos and @daN4cat I'm still working on purchasing but I've been informed by the guy funding my client's venture that he wants to develop his own system - I'll keep my opinions as to the cost/value benefit of his new mission to myself.

So my priorities have changed a bit.

I've really enjoyed working on this project and I'm going to continue to contribute but rather than working according to the priorities of my client I'll be working according to the priorities set by you (and as always as time permits).

So I'm tossing it out here as to whether I should temporarily redirect to GST Taxing or continue with Purchasing.

**Comment by daN4cat** [comment]
@SteveIreland what's the status of Purchasing work? Something done, nothing yet, else?

**Comment by daN4cat** [comment]
WRT GST, I would ask to be paid with a code bounty. People could take and share the cost as it's required for a specific place.

**Comment by daN4cat** [comment]
Either way, feel free to do what you prefer. In the end there is no strong commitment outside at least get a feature to be working to full quality.

**Comment by SteveIreland** [comment]
Okay.  As far as Purchasing I have the framework in place  I'm able to create,, update, and print a purchase order.  What is left to do is to flesh it out, put some checks and balances into it, and integrate it in with the other components.

**Comment by esbyrt** [comment]
When I originally set this up I was using simple tax. GST in Alberta is a flat 5% on everything except food items, which my users don't sell. I have no idea why so many of the inventory items have a tax code of zero but that seems to be the case with recently added items, as in after I updated to 3.1.1. I tested adding a new item now since I enabled customer origin tax code and the category is 1 when I check the database.

**Comment by SteveIreland** [comment]
I would recommend when switching midstream from simple sales tax (which does not support sales categories) to general sales tax you should set the tax category code on the item file to the the correct tax category  (Standard tax has a value of 1).

There is probably other data cleanup that may need to be looked at but since I don't know exactly what took place all I can do is point you in the general direction.  

If you changed from VAT to Sales tax then you have a real mess on your hands regarding tax reporting but that should never happen ... except maybe in India.

**Comment by SteveIreland** [comment]
@esbyrt was enough information provided that you were able to identify the issue?

**Comment by daN4cat** [comment]
Huston doesn't answer, so let's close.

**Comment by esbyrt** [comment]
Hi. I never did identify the problem other than an issue with taxes. To recap, after upgrading to 3.1.1 taxes didn't show up on the daily sales report and some sales would fail to complete. I set up a code under the new tax tab and the taxes were back on the reports. But, I still had an issue with two or three tax codes showing up with the sale and in those cases the sale would fail to complete. I found any items entered after the update had a tax code of 0. After further experimenting I enabled Customer Origin Tax Code in store config and updated all items through phpmyadmin to have the same tax codes. All seems to be working now so this can remain closed I guess.

**Comment by esbyrt** [comment]
Hi. I know this is closed but it cropped up again where two items with the same tax code showed on two separate tax lines in the sale register and after clicking complete the browser would go to blank screen instead of the receipt to print. Going back to the sale in the register it was still sitting there but just wouldn't enter.

I spend a few hours going over all the tax code settings and my inventory items and I think I've finally found the issue. My tax code on all items was GST but when I edited the item and just typed GST again and saved it would work. Sooo, long story short I discovered about 250 items in my inventory had two spaces after the GST code. It must have happened when I originally imported the items from a spreadsheets. I had done a search and replace for "GST" and "GST" which didn't fix the problem. When I searched for "GST  " and GST it found about 250 to change  (all without the quotes of course).  So for future reference I thought I'd share here in case someone else runs into this issue.
